### PR TITLE
Firebase auth

### PR DESCRIPTION
**Pull request by Alexdive**
Firebase auth
Auth manager
User screen
Adjusted UI and flow

### CODE REVIEW

File: LogIn.xcodeproj/xcuserdata/Aleksei_Permiakov.xcuserdatad/xcschemes/xcschememanagement.plist [hunk 1]
```diff
@@ -9,6 +9,27 @@
 			<key>orderHint</key>
 			<integer>0</integer>
```
Comment (0111b) [inline LogIn.xcodeproj/xcuserdata/Aleksei_Permiakov.xcuserdatad/xcschemes/xcschememanagement.plist]: you should avoid committing personal settings

File: LogIn.xcodeproj/xcuserdata/Aleksei_Permiakov.xcuserdatad/xcschemes/xcschememanagement.plist [hunk 2]
```diff
@@ -9,6 +9,27 @@
...
 		</dict>
+		<key>Promises (Playground) 1.xcscheme</key>
+		<dict>
+			<key>isShown</key>
+			<false/>
+			<key>orderHint</key>
+			<integer>2</integer>
+		</dict>
+		<key>Promises (Playground) 2.xcscheme</key>
```
Comment (0111b) [inline LogIn.xcodeproj/xcuserdata/Aleksei_Permiakov.xcuserdatad/xcschemes/xcschememanagement.plist]: What a heck is this?))

File: LogIn/LoginViewController/LoginViewModel.swift [hunk 3]
```diff
@@ -37,11 +40,20 @@ final class LoginViewModel: LoginViewModelType {
...
+    
     private func onLoginTap(email: String, passw: String) {
         if loginState == .login {
-            print("Log in", email, passw)
+            auth.signIn(email: email, password: passw, completion: completion)
         } else {
-            print("Sign up", email, passw)
+            auth.createUser(email: email, password: passw, completion: completion)
         }
     }
```
Comment (0111b) [inline LogIn/LoginViewController/LoginViewModel.swift:67]: use switch in the `onSignUpTap`

Comment (0111b) [inline LogIn/LoginViewController/LoginViewModel.swift:67]: Why do you need `emailString` and `passwString`?

File: LogIn/LoginViewController/LoginViewModel.swift [hunk 4]
```diff
@@ -15,17 +15,20 @@ enum LoginState {
 
 protocol LoginViewModelType {
     var presentationObject: LoginViewPresentationObject { get }
-    func transform(input: LoginViewModelInput) -> AnyPublisher<LoginViewModelOutput, Never>
     var transitionToLogin: (() -> Void)? { get set }
```
Comment (0111b) [inline LogIn/LoginViewController/LoginViewModel.swift]: looks like those 3 can be publishers

File: LogIn/UserViewController.swift [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+//  UserViewController.swift
+//  LogIn
+//
+//  Created by Aleksei Permiakov on 11.01.2022.
+//
+
+import UIKit
+
+final class UserViewController: UIViewController {
+
```
Comment (0111b) [inline LogIn/UserViewController.swift:11]: TODO: viewModel

File: LogIn/Auth/AuthManager.swift [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+//
+//  Created by Aleksei Permiakov on 11.01.2022.
+//
+
+import FirebaseAuth
+import Combine
+
+final class AuthManager {
+    
+    static let shared = AuthManager()
```
Comment (0111b) [inline LogIn/Auth/AuthManager.swift:13]: It will be hard to test with singletons everywhere

File: LogIn/Auth/AuthManager.swift [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    private init() {}
+    
+    var isLoggedIn = CurrentValueSubject<Bool, Never>(false)
```
Comment (0111b) [inline LogIn/Auth/AuthManager.swift]: Do not expose to the public writable interface. Anyone can change this right now

File: LogIn/Auth/AuthManager.swift [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+    
+    var userEmail: String? {
+        Auth.auth().currentUser?.email
+    }
+
```
Comment (0111b) [inline LogIn/Auth/AuthManager.swift:22]: Take a look https://www.donnywals.com/using-promises-and-futures-in-combine/ to see how to wrap completion API's with combine

**Review by 0111b — CHANGES_REQUESTED** [review]
Pretty nice but let's sync and discuss 😉

File: LogIn.xcodeproj/xcuserdata/Aleksei_Permiakov.xcuserdatad/xcschemes/xcschememanagement.plist (hunk 1 above)
Comment (Alexdive) [reply LogIn.xcodeproj/xcuserdata/Aleksei_Permiakov.xcuserdatad/xcschemes/xcschememanagement.plist]: I have no idea how it popped here, as xcuserdata folder is in gitignore

File: LogIn.xcodeproj/xcuserdata/Aleksei_Permiakov.xcuserdatad/xcschemes/xcschememanagement.plist (hunk 2 above)
Comment (Alexdive) [reply LogIn.xcodeproj/xcuserdata/Aleksei_Permiakov.xcuserdatad/xcschemes/xcschememanagement.plist]: hmmmm...

File: LogIn/Auth/AuthManager.swift (hunk 6 above)
Comment (Alexdive) [reply LogIn/Auth/AuthManager.swift:13]: As an option property can be added to init of the class where it been used

File: LogIn/Auth/AuthManager.swift (hunk 7 above)
Comment (Alexdive) [reply LogIn/Auth/AuthManager.swift]: ah yes

File: LogIn/LoginViewController/LoginViewModel.swift (hunk 3 above)
Comment (Alexdive) [reply LogIn/LoginViewController/LoginViewModel.swift:67]: to pass it in onLoginTap method

Comment (Alexdive) [reply LogIn/LoginViewController/LoginViewModel.swift:67]: Fixed

File: LogIn/LoginViewController/LoginViewModel.swift (hunk 4 above)
Comment (Alexdive) [reply LogIn/LoginViewController/LoginViewModel.swift]: True! Replaced

File: LogIn/Auth/AuthManager.swift [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+import FirebaseAuth
+import Combine
+
+final class AuthManager {
+    
+    static let shared = AuthManager()
+    private init() {}
+    
+    private(set) var isLoggedIn = CurrentValueSubject<Bool, Never>(false)
```
Comment (0111b) [inline LogIn/Auth/AuthManager.swift]: Not fixed, anyone can do isLoggedIn.send(false)

File: LogIn/LoginViewController/LoginViewController.swift [hunk 10]
```diff
@@ -161,12 +174,17 @@ final class LoginViewController: UIViewController {
...
+            .sink { _ in
+                self.transitionToLoginWithAnimation()
+            }
+            .store(in: &subscriptions)
+        
+        viewModel.transitionToSignUp
+            .sink { _ in
+                self.transitionToSignUpWithAnimation()
+            }
+            .store(in: &subscriptions)
```
Comment (0111b) [inline LogIn/LoginViewController/LoginViewController.swift:169]: Please double-check is there a retain cycle

**Review by 0111b — CHANGES_REQUESTED** [review]

**Pull request merged**